### PR TITLE
GEODE-4396: Separate Message.java into two concrete classes

### PR DESCRIPTION
**Pull request by afTrolle**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by bbaynes** [comment]
@bschuchardt @galen-pivotal @upthewaterspout @WireBaron @PivotalSarge

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java [hunk 1]
```diff
@@ -78,7 +78,7 @@
  *
  * @see MessageType
  */
-public class Message {
+public abstract class Message {
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java]: Is it possible to reduce the visibility of this class?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 2]
```diff
@@ -158,10 +158,10 @@ public static void emptyCommBufferPool() {
 
   // IMPORTANT: if new messages are added change setHandshake to initialize them
   // to the correct Version for serializing to the client
-  private Message requestMsg = new Message(2, Version.CURRENT);
-  private Message replyMsg = new Message(1, Version.CURRENT);
-  private Message responseMsg = new Message(1, Version.CURRENT);
-  private Message errorMsg = new Message(1, Version.CURRENT);
+  private Message requestMsg;
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: These could be changed to MessageFromClient & MessageFromServer.  I think that would improve the clarity of these declarations.  Then you could remove the casts to these classes that you've added in other parts of this file.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 3]
```diff
@@ -793,8 +807,8 @@ protected void doNormalMsg() {
             throw new AuthenticationRequiredException("Failed to find the authenticated user.");
           }
         }
-
-        command.execute(msg, this, this.securityService);
+        if (this.communicationMode.isCountedAsClientServerConnection())
```
Comment (bschuchardt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: I'm curious why you added this check.

**Comment by galen-pivotal** [comment]
I like this idea, but without some updates we can't merge it. @afTrolle are you interested in doing the work to get this mergeable?

**Comment by galen-pivotal** [comment]
Closing due to inactivity.

**Pull request closed without merging**